beluga_amcl/src/ndt_amcl_node_3d.cpp

```diff
@@ -0,0 +1,447 @@
+            execution_policy);
+      },
+      get_motion_model(), get_execution_policy());
+  return std::make_unique<NdtAmclVariant>(std::move(amcl));
+}
+
+void NdtAmclNode3D::do_initial_pose_callback(geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr message) {
+  auto pose = Sophus::SE3d{};
+  tf2::convert(message->pose.pose, pose);
+  const Eigen::Map<const Sophus::Matrix6d> covariance(message->pose.covariance.data());
+  last_known_estimate_ = std::make_pair(pose, covariance);
+  last_known_odom_transform_in_map_.reset();
+  initialize_from_estimate(last_known_estimate_.value());
+}
+
+void NdtAmclNode3D::do_periodic_timer_callback() {
+  if (!particle_filter_) {
+    return;
+  }
+
+  if (particle_cloud_pub_->get_subscription_count() == 0) {
+    return;
+  }
+  std::visit(
+      [this](const auto& particle_filter) {
+        auto message = beluga_ros::msg::PoseArray{};
+        beluga_ros::assign_particle_cloud(particle_filter.particles(), message);
+        beluga_ros::stamp_message(get_parameter("global_frame_id").as_string(), now(), message);
+        particle_cloud_pub_->publish(message);
+      },
+      *particle_filter_);
+}
+
+// TODO(alon): Wouldn't it be better in the callback of each message to simply receive
+// it and define another timer or thread to do the work of calculation and publication?
+void NdtAmclNode3D::laser_callback(sensor_msgs::msg::PointCloud2::ConstSharedPtr laser_scan) {
+  if (!particle_filter_ || !last_known_estimate_.has_value()) {
+    RCLCPP_WARN_THROTTLE(
+        get_logger(), *get_clock(), 2000, "Ignoring laser data because the particle filter has not been initialized");
+    return;
+  }
+
+  auto base_pose_in_odom = Sophus::SE3d{};
+  auto laser_pose_in_base = Sophus::SE3d{};
+  try {
+    // Use the lookupTransform overload with no timeout since we're not using a dedicated
+    // tf thread. The message filter we are using avoids the need for it.
+    tf2::convert(
+        tf_buffer_
+            ->lookupTransform(
+                get_parameter("odom_frame_id").as_string(), get_parameter("base_frame_id").as_string(),
+                tf2_ros::fromMsg(laser_scan->header.stamp))
+            .transform,
+        base_pose_in_odom);
+    tf2::convert(
+        tf_buffer_
+            ->lookupTransform(
+                get_parameter("base_frame_id").as_string(), laser_scan->header.frame_id,
+                tf2_ros::fromMsg(laser_scan->header.stamp))
+            .transform,
+        laser_pose_in_base);
+  } catch (const tf2::TransformException& error) {
+    RCLCPP_ERROR(get_logger(), "Couldn't find needed transform, : %s", error.what());
+    return;
+  }
+
+  const auto update_start_time = std::chrono::high_resolution_clock::now();
+
+  std::vector<Eigen::Vector3d> measurement;
+  measurement.reserve(laser_scan->height * laser_scan->width);
+
+  //  Accessing XYZ as suggested here:
+  // https://docs.ros.org/en/jade/api/sensor_msgs/html/classsensor__msgs_1_1PointCloud2Iterator.html
```

Comment:
@serraramiro1 meta: this makes extra assumptions about pointcloud field layout. IMO an iterator for each well-known named field does not make such assumptions.

Reply:
Addressed, cannot use fancy ranges things as these iterators are custom classes that don't adhere to iterator conventions.
(No `begin()`, for instance).